### PR TITLE
Object safe for dispatch

### PR DESCRIPTION
**Pull request by bovinebuddha**
cc #43561

### CODE REVIEW

**Comment by nikomatsakis** [comment]
@bovinebuddha I'm sorry for not getting back to you in all this time! It's been a very busy couple of weeks. I've added it to my calendar to review on Monday Jan 28, so hopefully that will help.

**Comment by bovinebuddha** [comment]
@nikomatsakis It's alright, I get gather you've got quite a full schedule. Just excited that its finally getting reviewed.

File: src/librustc_typeck/check/coercion.rs
Comment (nikomatsakis) [inline src/librustc_typeck/check/coercion.rs]: Note to self: double check how these traits work, do we always want to apply this test here?

File: src/librustc/ty/wf.rs
Comment (nikomatsakis) [inline src/librustc/ty/wf.rs]: Note to self: think over the interaction of implied bounds here

**Review by nikomatsakis — COMMENTED** [review]
OK, I did a first pass. The code seems very nice.

I left myself some notes for next time. I want to come back and re-read the RFC a bit. I also wanted to reconsider some potential interactions with other parts of the trait system.

(I've scheduled a slot in my calendar for this on this coming Friday.)

**Comment by Dylan-DPC** [comment]
ping from triage @bovinebuddha you have conflicts to resolve

File: src/librustc/traits/select.rs
Comment (mikeyhew) [inline src/librustc/traits/select.rs]: Is the check for the feature flag necessary here? It seems like you could check if the trait is object safe anyway, even if the feature is not enabled. Or would that result in duplicate errors?

**Comment by mikeyhew** [comment]
About the feature name: I know it comes from the RFC, but to me the feature is more general than just allowing static dispatch on `dyn Trait` types. The feature is that in order for `dyn Trait` to be a well-formed type, `Trait` no longer needs to be object-safe. Maybe a name like `dyn_trait_always_well_formed` would get that across.

File: src/librustc/traits/select.rs
Comment (bovinebuddha) [reply src/librustc/traits/select.rs]: I believe this is the code which adds the automatic 'impl Trait for dyn Trait' predicate to trait objects. Maybe the branches should be clarified and a comment added. Like:

// Add comments here
let object_impl_trait = !self.infcx.tcx.features().object_safe_for_dispatch ||
                                    self.tcx().is_object_safe(principal.def_id());

if (object_impl_trait) {
  principal.with_self_ty(self.tcx(), self_ty)
} else {
 return;
}

**Comment by bovinebuddha** [comment]
I also didn't really like using the RFC name for the  feature. I suggested something like 'symmetric_object_safety' but I think your name might be more to the point?

**Comment by bovinebuddha** [comment]
@Dylan-DPC, could you please add back S-waiting-on-review?

**Comment by Dylan-DPC** [comment]
@nikomatsakis you can review this now (before new conflicts arise 😄 )

**Comment by bovinebuddha** [comment]
@Dylan-DPC, @nikomatsakis here we go again :)

File: src/libsyntax/feature_gate.rs
Comment (Centril) [inline src/libsyntax/feature_gate.rs]:
```suggestion
    (active, object_safe_for_dispatch, "1.35.0", Some(43561), None),
```

File: src/test/ui/feature-gates/feature-gate-object_safe_for_dispatch.stderr
Comment (Centril) [inline src/test/ui/feature-gates/feature-gate-object_safe_for_dispatch.stderr]: Seems the feature gate is not suggested?

Comment (bovinebuddha) [reply src/test/ui/feature-gates/feature-gate-object_safe_for_dispatch.stderr]: Thanks, I'll try and fix that. Although, I guess such suggestions should not appear on stable. Is there any way to distinguish between stable and nightly in the code? Or are those separate branches?

Comment (mikeyhew) [reply src/test/ui/feature-gates/feature-gate-object_safe_for_dispatch.stderr]: If you use `syntax::feature_gate::feature_err`, it will only suggest the feature on nightly

**Comment by Dylan-DPC** [comment]
ping from triage @nikomatsakis waiting for your review on this

**Comment by nikomatsakis** [comment]
I'm sorry @bovinebuddha for the unforgiveable delays here. I've scheduled this for review this coming Thu, 18 Apr, at 13:00 UTC-04:00.

**Comment by nikomatsakis** [comment]
Argh. I got overly busy today. Rescheduled to Friday at 12:00 UTC-04:00.

File: src/librustc/traits/select.rs
Comment (nikomatsakis) [reply src/librustc/traits/select.rs]: At minimum, a comment would be good here -- I think I vaguely prefer the `if object_impl_trait { .. }` version as well.

But a comment like this would be great:

---

// Prior to RFC 2027, `dyn Foo: Foo` was always true. But RFC 2027 makes that only true if `Foo` is object safe.

File: src/librustc/ty/wf.rs
Comment (nikomatsakis) [reply src/librustc/ty/wf.rs]: So I haven't really managed to think this over very much, though I did [ping @scalexm over on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/144729-wg-traits/topic/object.20safe.20for.20dispatch.20.2357545) to start a conversation there (and I'd rather have that conversation there, or at least not on this PR). That said, I guess that we can land this PR, but we might have to roll it back.

File: src/librustc_errors/diagnostic_builder.rs
Comment (nikomatsakis) [inline src/librustc_errors/diagnostic_builder.rs]: Nit: I think it is spelled `unsuccessful` (with one `l`), at least in American English.

File: src/librustc_typeck/check/coercion.rs
Comment (nikomatsakis) [inline src/librustc_typeck/check/coercion.rs]: Ah, I guess this is the code I was recommending we delete (I still think we should delete it) -- but we could achieve something similar to this commit I suspect by adjusting the "cause" we wind up with.

File: src/librustc_typeck/check/cast.rs
Comment (nikomatsakis) [inline src/librustc_typeck/check/cast.rs]: I think these changes would fall out from the approach I highlighted, and then changing the trait error handling code.

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
OK. I'm feeling pretty good about this PR. I made one suggested change. Let me know what you think -- I suspect after we resolve this question, we'd be ready to land.

**Comment by bovinebuddha** [comment]
@nikomatsakis

Thanks for the review! Been on vacation for a while, coming back next week and try and work then through.

**Comment by Centril** [comment]
Ping from triage. @bovinebuddha Back from your vacation yet? :)

**Comment by bovinebuddha** [comment]
@nikomatsakis Ok, let's tackle this.

So. I think I understand what you're getting at, I'm gonna try and revert all the code in coerion.rs and simply leave it to the later candidate confirmation.

However, this is gonna screw with the diagostics I think, so it's gonna take some time to wrap my head around. It's been some time since I wrote this code, gonna have to dust of my brain and penetrate the rust source code again.

So I'll revert the suggested code and run the UI tests and see what happens.

File: src/librustc_typeck/check/coercion.rs
Comment (bovinebuddha) [reply src/librustc_typeck/check/coercion.rs]: Ok so deep diving into my memory, I think what I was getting at was trying to turn change the error message: 

```rust
 LL |     Box::new(S) as Box<dyn Trait>;
    |     ^^^^^^^^^^^ the trait `Trait` cannot be made into an object
    |
    = note: the trait cannot require that `Self : Sized`
    = note: required because of the requirements on the impl of `std::ops::CoerceUnsized<std::boxed::Box<dyn Trait>>` for `std::boxed::Box<S>`
```
into the following:

```rust
 LL |     Box::new(S) as Box<dyn Trait>;
    |                    ^^^^^^^^^^^^^^ the trait `Trait` cannot be made into an object
    |
    = note: the trait cannot require that `Self : Sized`
    = note: required by cast to type 'std::boxed::Box<dyn Trait>'
```

I though that trying to coerce a struct into a non-object-safe trait will be quite a common occurance, and the first error message will serve as a strong deterrent to sanity for new adopters. I think the new error message is better for two reasons:

1. Obviously, there is no mention of CoerceUnsized.
2. The error span marks the 'trying to construct a non-object safe trait object' part of the code.

I think that question 1 could probably be resolved by reworking my diagnostics commit and getting the information from trait selection instead, but I'm not sure if the desired span survives is available at that point. Will have to do some more digging for that.

**Comment by bovinebuddha** [comment]
@nikomatsakis Ok, let's tackle this.

So. I think I understand what you're getting at, I'm gonna try and revert all the code in coerion.rs and simply leave it to the later candidate confirmation.

However, this is gonna screw with the diagostics I think, so it's gonna take some time to wrap my head around. It's been some time since I wrote this code, gonna have to dust of my brain and penetrate the rust source code again.

So I'll revert the suggested code and run the UI tests and see what happens.

**Comment by bovinebuddha** [comment]
@nikomatsakis 
I answered your review question.

If you think the error message is a non-blocker, I can probably do as you asked and just remove the code, along with most of the diagnostics commit. Maybe add somewhere that error messages should be reviewed before stabilization.

**Comment by nikomatsakis** [comment]
@bovinebuddha 

Sorry for not responding again -- I was traveling, but I'm back now.

> I though that trying to coerce a struct into a non-object-safe trait will be quite a common occurance, and the first error message will serve as a strong deterrent to sanity for new adopters.

I like this phrasing. =) I do agree the error is ungreat, but I also think I'd rather try and address that in follow-up PRs. 

However, I don't actually see that you pushed any commits reverting the change? 

I do prefer to try to keep the "should we coerce logic" as clean as possible, and tackle the error reporting separately.

**Comment by Dylan-DPC** [comment]
ping from triage @bovinebuddha any updates?

**Comment by nikomatsakis** [comment]
Given that @bovinebuddha hasn't been responding, and how long this PR has been open (eek), I took the liberty of rebasing their branch. I'm running tests now.

**Comment by rholderfield** [comment]
ping from triage... hi @nikomatsakis, are you still working on this? if so, can you please provide a status?

**Comment by Dylan-DPC** [comment]
Closing this PR due to inactivity

**Comment by nikomatsakis** [comment]
I'm reopening this -- sorry for lack of response. I'm on an extended vacation, but I'll be back next week and will try to decide whether to land this PR or what. I did rebase it and seemed like it was passing tests, and I would be able to rebase again, though obviously the history needs to be cleaned up, too.

**Comment by JohnCSimon** [comment]
Ping from triage -
@nikomatsakis 
This PR has a merge conflict that needs to be resolved, but after that it should be ready for review
Putting this back into the waiting-on-author pile.
Thank you!

**Comment by JohnCSimon** [comment]
Ping from triage -
@nikomatsakis how is your vacation? :)

**Comment by JohnTitor** [comment]
Ping from triage: @nikomatsakis any updates on this?

**Comment by Dylan-DPC** [comment]
@bovinebuddha needs to resolve conflicts

**Comment by nikomatsakis** [comment]
Update: have rebased this and I pushed a commit that addressed my concern, but I have to decide if I want to make any changes related to diagnostics. Maybe at this point I should just close it and open a new PR?

**Comment by JohnCSimon** [comment]
Ping from triage 
@bovinebuddha @nikomatsakis This has sat idle for nine days and there's a conflict. Do you want to go ahead and close this PR and open another one?
Thanks.

**Comment by nikomatsakis** [comment]
OK so I just realized that the impact on diagnostics is limited to when the feature gate is enabled -- also, I don't think it's that bad, and I have a tentative PR that will make improvements. But I'm inclined to land that work separately. So I see no reason not to land this PR.

**Comment by nikomatsakis** [comment]
Hmm, however, our experiences working on fixing #57983 suggest that enabling this feature gate in libcore and friends may create a cycle error.

**Comment by nikomatsakis** [comment]
But I think we should land this anyway.

**Comment by nikomatsakis** [comment]
Also, I pushed an improvement to diagnostics.

**Comment by Centril** [comment]
(When you do land this, please also add an F- label to the PR and the tracking issue. // triage)

**Comment by JohnCSimon** [comment]
Ping from triage
@bovinebuddha can you please address the merge conflict?
@Centril @nikomatsakis 

Thank you!

**Comment by nikomatsakis** [comment]
@bors r-

Ugh the history is so terrible I can't stand it. Will squash.

File: src/librustc/traits/select.rs
Comment (Centril) [inline src/librustc/traits/select.rs]: `else if`?

**Comment by joelpalmer** [comment]
Ping from Triage: any update @bovinebuddha?

**Comment by nikomatsakis** [comment]
I'm not sure what the problem is here-- I see no problem locally

**Comment by nikomatsakis** [comment]
@bors r+ 

I'm going to guess that this error is spurious somehow.

**Comment by Centril** [comment]
Using this as a canary to check infra status; @bors p=1001

File: src/test/ui/feature-gates/feature-gate-object_safe_for_dispatch.stderr
Comment (Centril) [reply src/test/ui/feature-gates/feature-gate-object_safe_for_dispatch.stderr]: @nikomatsakis This was never fixed :(

**Comment by mikeyhew** [comment]
@nikomatsakis once this is stabilized, we remove this hacky code from object_safety.rs

https://github.com/rust-lang/rust/blob/8318ef26c1ad949636a161437ec251df59cf514a/src/librustc/traits/object_safety.rs#L558

In that code, we substitute in `U` for `Self`, and add `Self: Unsize<U>` to the param env. Instead, we'll be able to substitute `dyn Trait` in for `Self`. We can't do that right now, because using `dyn Trait` in the predicate obligation triggers another check that `Trait` is object safe, and we get a query cycle.

**Comment by Centril** [comment]
@mikeyhew Can you extend the FIXME?

**Comment by mikeyhew** [comment]
@Centril good call. I've opened #65850 to address that

**Pull request merged**
